Tests/NIOTests/SocketChannelTest.swift

```diff
@@ -297,4 +297,82 @@ public class SocketChannelTest : XCTestCase {
             XCTFail("Unexpected error \(error)")
         }
     }
+
+    public func testPendingConnectNotificationOrder() throws {
+        class ConnectPendingSocket: Socket {
+            init() throws {
+                try super.init(protocolFamily: PF_INET, type: Posix.SOCK_STREAM)
+            }
+
+            override func connect(to address: SocketAddress) throws -> Bool {
+                return false
+            }
+        }
+
+        class NotificationOrderHandler: ChannelDuplexHandler {
+            typealias InboundIn = Any
+            typealias OutboundIn = Any
+
+            private var connectPromise: EventLoopPromise<Void>?
+
+            public func channelInactive(ctx: ChannelHandlerContext) {
+                if let connectPromise = self.connectPromise {
+                    XCTAssertTrue(connectPromise.futureResult.isFulfilled)
+                } else {
+                    XCTFail("connect(...) not called before")
+                }
+            }
+
+            public func connect(ctx: ChannelHandlerContext, to address: SocketAddress, promise: EventLoopPromise<Void>?) {
+                XCTAssertNil(self.connectPromise)
+                self.connectPromise = promise
+                ctx.connect(to: address, promise: promise)
+            }
+
+            func handlerAdded(ctx: ChannelHandlerContext) {
+                XCTAssertNil(self.connectPromise)
+            }
+
+            func handlerRemoved(ctx: ChannelHandlerContext) {
+                if let connectPromise = self.connectPromise {
+                    XCTAssertTrue(connectPromise.futureResult.isFulfilled)
+                } else {
+                    XCTFail("connect(...) not called before")
+                }
+            }
+        }
+
+        let group = MultiThreadedEventLoopGroup(numThreads: 1)
+        defer { XCTAssertNoThrow(try group.syncShutdownGracefully()) }
+
+        let channel = try SocketChannel(eventLoop: group.next() as! SelectableEventLoop, protocolFamily: PF_INET)
+
+        XCTAssertNoThrow(try channel.pipeline.add(handler: NotificationOrderHandler()).then {
+            channel.register()
+        }.wait())
+
+        let connectPromise: EventLoopPromise<Void> = channel.eventLoop.newPromise()
+        let closePromise: EventLoopPromise<Void> = channel.eventLoop.newPromise()
+
+        channel.connect(to: try! SocketAddress.init(ipAddress: "127.0.0.1", port: 9999), promise: connectPromise)
```

Comment:
extra `.init`

Reply:
move `try` to the beginning of the line and add `XCTAssertNoThrow(...)`

Reply:
@weissi I just removed the `!` which should be good enough as it will fail the test.